File: 6-zigzag-conversion/6-zigzag-conversion.py
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        if numRows == 1 or numRows >= len(s):
            return s
        
        rows = [[] for _ in range(numRows)]
        
        r, dr = 0, -1
        
        for c in s:
            rows[r].append(c)
            
            if r == 0 or r == numRows - 1:
                dr = -dr
            
            r += dr
        
        return ''.join(''.join(row) for row in rows)

File: 6-zigzag-conversion/6-zigzag-conversion.go
func convert(s string, numRows int) string {
    if numRows == 1 || numRows >= len(s) {
        return s
    }
    
    rows := make([]strings.Builder, numRows)
    
    r, dr := 0, -1
    
    for _, c := range s {
        rows[r].WriteRune(c)
        
        if r == 0 || r == numRows - 1 {
            dr = -dr
        }
        
        r += dr
    }
    
    strs := make([]string, numRows)
    for i, row := range rows {
        strs[i] = row.String()
    }
    return strings.Join(strs, "")
}